### ISSUE TITLE
What is supposed to happen when choosing option 2 (MSYS2 repository update) from ridk?

### ISSUE THREAD

**Issue by LamentConfiguration** [oneclick/rubyinstaller2#182]
Is there more documentation for using the ridk tool? I generally tend to run an update after installing, and I'm not sure that option 2 is happening the way it should be, as it is asking where I want to install the MSYS2 components, when I choose the folder initially created by the 2.7.1 devkit installer, it asks me if I want to over-write the current files, and this doesn't seem correct to me when I would choose an "update" option.

**Comment by larskanis** [comment]
Sorry you having problems to install ruby! There's some documentation to `ridk` [in the wiki](https://github.com/oneclick/rubyinstaller2/wiki/The-ridk-tool).

Unfortunately I'm uncertain if I understand you right. Could you please copy and past some relevant outputs? You can also copy and paste screenshots easily into github issues.

**Comment by LamentConfiguration** [comment]
I will attempt it all from scratch again, and take better notes/screenshots (i don't have any problem uninstalling and reinstalling) when I get a chance in the next couple days. Thank you in advance for your assistance.

**Comment by LamentConfiguration** [comment]
So, I'm trying again! I rand the 2.7.1 with devkit installer, and it installed just fine, the ridk options popped up in a command window at the end, and i ran the default of 1,3, and it went fine.

Here's the output from ridk version afterwards:
D:\rubies>ridk version
---
ruby:
  path: D:/rubies/Ruby27-x64
  version: 2.7.1
  platform: x64-mingw32
ruby_installer:
  package_version: 2.7.1-1
  git_commit: eb60b1c
msys2:
  path: D:\rubies\Ruby27-x64\msys64
cc: x86_64-w64-mingw32-gcc (Rev1, Built by MSYS2 project) 9.3.0
sh: GNU bash, version 4.4.23(2)-release (x86_64-pc-msys)
os: Microsoft Windows [Version 10.0.18362.959]

I then ran ridk install, and selected option 2
and get this:

![image](https://user-images.githubusercontent.com/15186074/88486231-e630c080-cf41-11ea-8895-1dff50c17b7f.png)


I'm not sure what I've done wrong, or if this isn't wrong, or what to do to make it right if it is wrong.

Thanks for all the help!

**Comment by kou** [comment]
This'll be fixed by updating `msys2-keyring` package manually before updating `pacman`: https://www.msys2.org/news/#2020-06-29-new-packagers

The following patch may fix this but I didn't test it...:

```diff
diff --git a/lib/ruby_installer/build/components/01_msys2.rb b/lib/ruby_installer/build/components/01_msys2.rb
index f3f0a68..a7fdddf 100644
--- a/lib/ruby_installer/build/components/01_msys2.rb
+++ b/lib/ruby_installer/build/components/01_msys2.rb
@@ -23,33 +23,15 @@ class Msys2 < Base
   def execute(args)
     require "open-uri"
 
-    uri = msys2_download_uri
-    filename = File.basename(uri)
-    temp_path = File.join(ENV["TMP"] || ENV["TEMP"] || ENV["USERPROFILE"] || "C:/", filename)
-
-    until check_hash(temp_path, msys2_download_hash)
-      puts "Download #{yellow(uri)}\n  to #{yellow(temp_path)}"
-      File.open(temp_path, "wb") do |fd|
-        progress = 0
-        total = 0
-        params = {
-          "Accept-Encoding" => 'identity',
-          :content_length_proc => lambda{|length| total = length },
-          :progress_proc => lambda{|bytes|
-            new_progress = (bytes * 100) / total
-            print "\rDownloading %s (%3d%%) " % [filename, new_progress]
-            progress = new_progress
-          }
-        }
-        OpenURI.open_uri(uri, params) do |io|
-          fd << io.read
-        end
-        puts
-      end
+    if ENV['MSYS2_VERSION']
+      hash = nil
+    else
+      hash = msys2_download_hash
     end
+    downloaded_path = download(msys2_download_uri, hash)
 
     puts "Run the MSYS2 installer ..."
-    if run_verbose(temp_path) && msys.with_msys_apps_enabled { run_verbose("sh", "-lc", "true") }
+    if run_verbose(downloaded_path) && msys.with_msys_apps_enabled { run_verbose("sh", "-lc", "true") }
       puts green(" Success")
     else
       puts red(" Failed")
@@ -59,6 +41,8 @@ class Msys2 < Base
 
   private
 
+  # We should remove "Updating MSYS2 keyring ..." section from
+  # 02_pacman_update.rb when we update MSYS2 version.
   MSYS2_VERSION = ENV['MSYS2_VERSION'] || "20190524"
   MSYS2_URI = "http://repo.msys2.org/distrib/<arch>/msys2-<arch>-#{MSYS2_VERSION}.exe"
 
@@ -76,21 +60,6 @@ class Msys2 < Base
       else MSYS2_I686_SHA256
     end
   end
-
-  def check_hash(path, hash)
-    if ENV['MSYS2_VERSION']
-      true
-    elsif !File.exist?(path)
-      false
-    else
-      require "digest"
-
-      print "Verify integrity of #{File.basename(path)} ..."
-      res = Digest::SHA256.file(path).hexdigest == hash.downcase
-      puts(res ? green(" OK") : red(" Failed"))
-      res
-    end
-  end
 end
 end
 end
diff --git a/lib/ruby_installer/build/components/02_pacman_update.rb b/lib/ruby_installer/build/components/02_pacman_update.rb
index 6f1ad17..837f780 100644
--- a/lib/ruby_installer/build/components/02_pacman_update.rb
+++ b/lib/ruby_installer/build/components/02_pacman_update.rb
@@ -12,6 +12,16 @@ class PacmanUpdate < Base
 
   def execute(args)
     msys.with_msys_apps_enabled do
+      # We need to update MSYS2 keyring manually when we use old MSYS2
+      # installer. See https://www.msys2.org/news/#2020-06-29-new-packagers
+      # for details.
+      # We can remove this when we update MSYS2 installer version.
+      puts "Updating MSYS2 keyring ..."
+      msys2_keyring_url = "https://repo.msys2.org/msys/x86_64/msys2-keyring-r21.b39fb11-1-any.pkg.tar.xz"
+      msys2_keyring_sha256 = "https://repo.msys2.org/msys/x86_64/msys2-keyring-r21.b39fb11-1-any.pkg.tar.xz"
+      msys2_keyring_path = download(msys2_keyring_url, msys2_keyring_sha256)
+      run_verbose("pacman", "-U", "--noconfirm", msys2_keyring_path)
+
       puts "Remove catgets to avoid conflicts while update  ..."
       # See https://github.com/Alexpux/MSYS2-packages/issues/1141
       run_verbose("pacman", "-Rdd", "catgets", "libcatgets", "--noconfirm")
diff --git a/lib/ruby_installer/build/components/base.rb b/lib/ruby_installer/build/components/base.rb
index 588a1b4..26b091c 100644
--- a/lib/ruby_installer/build/components/base.rb
+++ b/lib/ruby_installer/build/components/base.rb
@@ -52,6 +52,49 @@ class Base < Rake::Task
   def puts(*args)
     $stderr.puts *args
   end
+
+  def download(url, hash)
+    require "open-uri"
+
+    filename = File.basename(uri)
+    temp_path = File.join(ENV["TMP"] || ENV["TEMP"] || ENV["USERPROFILE"] || "C:/", filename)
+    until check_hash(temp_path, hash)
+      puts "Download #{yellow(uri)}\n  to #{yellow(temp_path)}"
+      File.open(temp_path, "wb") do |fd|
+        progress = 0
+        total = 0
+        params = {
+          "Accept-Encoding" => 'identity',
+          :content_length_proc => lambda{|length| total = length },
+          :progress_proc => lambda{|bytes|
+            new_progress = (bytes * 100) / total
+            print "\rDownloading %s (%3d%%) " % [filename, new_progress]
+            progress = new_progress
+          }
+        }
+        OpenURI.open_uri(uri, params) do |io|
+          fd << io.read
+        end
+        puts
+      end
+    end
+    temp_path
+  end
+
+  def check_hash(path, hash)
+    if !File.exist?(path)
+      false
+    elsif hash.nil?
+      false
+    else
+      require "digest"
+
+      print "Verify integrity of #{File.basename(path)} ..."
+      res = Digest::SHA256.file(path).hexdigest == hash.downcase
+      puts(res ? green(" OK") : red(" Failed"))
+      res
+    end
+  end
 end
 end
 end
```

Other solution is updating MSYS2 installer:

```diff
diff --git a/lib/ruby_installer/build/components/01_msys2.rb b/lib/ruby_installer/build/components/01_msys2.rb
index f3f0a68..b26bf11 100644
--- a/lib/ruby_installer/build/components/01_msys2.rb
+++ b/lib/ruby_installer/build/components/01_msys2.rb
@@ -59,11 +59,11 @@ class Msys2 < Base
 
   private
 
-  MSYS2_VERSION = ENV['MSYS2_VERSION'] || "20190524"
+  MSYS2_VERSION = ENV['MSYS2_VERSION'] || "20200720"
   MSYS2_URI = "http://repo.msys2.org/distrib/<arch>/msys2-<arch>-#{MSYS2_VERSION}.exe"
 
   MSYS2_I686_SHA256 = "c5a1881f7ac5a0449fe9b30d3140111a88072727f510c4a66bfa905f8c78e839"
-  MSYS2_X86_64_SHA256 = "2dacadcc70cc122054e60914cbc6b689f685bef5713915a90f4185dd9da7954e"
+  MSYS2_X86_64_SHA256 = "ad6b58b355a852d9df7c6dc2722d0eca139042cf262ae7363e0a5e7dd2a6c5c9"
 
   def msys2_download_uri
     arch = RUBY_PLATFORM=~/x64/ ? "x86_64" : "i686"
```

But it drops support for i686 because MSYS2 installer for i686 is no longer released: https://www.msys2.org/news/#2020-05-17-32-bit-msys2-no-longer-actively-supported

**Comment by MSP-Greg** [comment]
Base install version are now also available at:
https://github.com/msys2/msys2-installer/tags

**Comment by larskanis** [comment]
Sorry for the slow response! This issue should be fixed by commit 29f4d3a51ca5660b47ed8143cba22306908a2d70 . You can update msys2 either manually by the steps of commit a7b2d950f2790406a6c17b9e951ae4712acbbd77 or by using `ridk install 1 2 3` from the latest rubyinstaller builds. They are available in the build artifacts section for various ruby versions here: https://ci.appveyor.com/project/larskanis/rubyinstaller2-hbuor/builds/34807789

**Comment by larskanis** [comment]
Thank you @kou for the patch! I changed the code very similar.

Regarding the future of the i686 build I opened #185.
